fix(core): fix to prevent policies from being applied to containers t…

haytok opened the pull request:
…hat shouldn't be applied

In running KubeArmor in systemd mode, when the container name (the value of `kubearmor.io/container.name`) in the applied policy was changed and the policy was applied, there was an issue where that policy was applied not only to the new container but also to the previous container.

This bug was also reported in issue/#1387.

Therefore, in this commit, the policy doesn't apply to the previous container, and apply to the new container.

**Purpose of PR?**:

Fixes #1387

**Does this PR introduce a breaking change?**

No

**If the changes in this PR are manually verified, list down the scenarios covered:**:

Based on the two tests below, I'll  show that this commit has been made to prevent the policy from being applied to containers that should not have been applied. 

### Note

I confirmed that the bug has been fixed by writing a debug program to check `dm.Endpoints` in the following part of [parseAndUpdateContainerSecurityPolicy()](https://github.com/kubearmor/KubeArmor/blob/main/KubeArmor/core/unorchestratedUpdates.go#L83C1-L546C2). 

<!--
ParseAndUpdateContainerSecurityPolicy() の下記の箇所に dm.EndPoints を確認するデバッグプログラムを記述して不具合が修正されていることを確認しました。
-->

https://github.com/kubearmor/KubeArmor/blob/main/KubeArmor/core/unorchestratedUpdates.go#L545

<details>
<summary> Debug code </summary>

```golang
ubuntu ~/KubeArmor/KubeArmor [fixed-not-to-overwrite-policies-already-applied]
> git diff core/unorchestratedUpdates.go
diff --git a/KubeArmor/core/unorchestratedUpdates.go b/KubeArmor/core/unorchestratedUpdates.go
index 1e87c341..b78c562d 100644
--- a/KubeArmor/core/unorchestratedUpdates.go
+++ b/KubeArmor/core/unorchestratedUpdates.go
@@ -5,6 +5,7 @@ package core

 import (
        "encoding/json"
+       "fmt"
        "os"
        "sort"
        "strings"
@@ -564,6 +565,10 @@ func (dm *KubeArmorDaemon) ParseAndUpdateContainerSecurityPolicy(event tp.K8sKub
                        dm.removeBackUpPolicy(secPolicy.Metadata["policyName"])
                }
        }
+
+       for i, endPoint := range dm.EndPoints {
+               fmt.Printf("i: %d, EndPointName: %s, number of policies: %d\n", i, endPoint.EndPointName, len(endPoint.SecurityPolicies))
+       }
 }

 // ================================= //
```

</details>

### Test case 1

1. Create `block_diff.yaml`

<details>
<summary> block_diff.yaml  </summary>

```yaml
ubuntu ~
> cat block_diff.yaml
apiVersion: security.kubearmor.com/v1
kind: KubeArmorPolicy
metadata:
  name: block-diff
spec:
  selector:
    matchLabels:
      # policyName
      kubearmor.io/container.name: hoge
  process:
    matchPaths:
      - path: /usr/bin/diff
  action:
    Block
```

</details>

2. apply `block_diff.yaml` by runnig `karmor vm policy add block_diff.yaml`

3. `dm.EndPoints` from debug message are shown below.

```bash
Sep 09 05:14:09 ip-172-31-33-167 kubearmor[377347]: 2023-09-09 05:14:09.880226        INFO        Detected a Container Security Policy (added/container_namespace/block-diff)
Sep 09 05:14:09 ip-172-31-33-167 kubearmor[377347]: 2023-09-09 05:14:09.903867        INFO        Registered the AppArmor profile (kubearmor_hoge)
Sep 09 05:14:09 ip-172-31-33-167 kubearmor[377347]: 2023-09-09 05:14:09.929617        INFO        Updated 1 security rule(s) to container_namespace/hoge/kubearmor_hoge
Sep 09 05:14:09 ip-172-31-33-167 kubearmor[377347]: i: 0, EndPointName: hoge, number of policies: 1
```

4. Update below `kubearmor.io/container.name` of `block_diff.yaml` and apply the policy

<details>
<summary> block_diff.yaml  </summary>

```yaml
ubuntu ~
> cat block_diff.yaml
apiVersion: security.kubearmor.com/v1
kind: KubeArmorPolicy
metadata:
  name: block-diff
spec:
  selector:
    matchLabels:
      # policyName
      kubearmor.io/container.name: fuga
  process:
    matchPaths:
      - path: /usr/bin/diff
  action:
    Block
```

</details>

5. `dm.EndPoints` are shown below.

```bash
Sep 09 05:14:37 ip-172-31-33-167 kubearmor[377347]: 2023-09-09 05:14:37.767391        INFO        Detected a Container Security Policy (added/container_namespace/block-diff)
Sep 09 05:14:37 ip-172-31-33-167 kubearmor[377347]: 2023-09-09 05:14:37.791037        INFO        Registered the AppArmor profile (kubearmor_fuga)
Sep 09 05:14:37 ip-172-31-33-167 kubearmor[377347]: 2023-09-09 05:14:37.816894        INFO        Updated 1 security rule(s) to container_namespace/fuga/kubearmor_fuga
Sep 09 05:14:37 ip-172-31-33-167 kubearmor[377347]: i: 0, EndPointName: fuga, number of policies: 1
```

From results of 5, the policy doesn't apply to the previous container, and apply to the new container.

<!--
なお、このコミットによる修正が変更されていない場合に上述の流れで policy を apply すると下記のようなデバッグメッセージが出力されます。
-->

Note that when the modifications due to this commit have not been changed, the following debug message will be output when the policy is applied in the above flow. 

```bash
Sep 09 05:10:05 ip-172-31-33-167 kubearmor[376347]: 2023-09-09 05:10:05.317470        INFO        Detected a Container Security Policy (added/container_namespace/block-diff)
Sep 09 05:10:05 ip-172-31-33-167 kubearmor[376347]: 2023-09-09 05:10:05.341019        INFO        Registered the AppArmor profile (kubearmor_hoge)
Sep 09 05:10:05 ip-172-31-33-167 kubearmor[376347]: 2023-09-09 05:10:05.367571        INFO        Updated 1 security rule(s) to container_namespace/hoge/kubearmor_hoge
Sep 09 05:10:05 ip-172-31-33-167 kubearmor[376347]: i: 0, EndPointName: hoge, number of policies: 1

# update container name
Sep 09 05:10:21 ip-172-31-33-167 kubearmor[376347]: 2023-09-09 05:10:21.261441        INFO        Detected a Container Security Policy (added/container_namespace/block-diff)
Sep 09 05:10:21 ip-172-31-33-167 kubearmor[376347]: 2023-09-09 05:10:21.284971        INFO        Registered the AppArmor profile (kubearmor_fuga)
Sep 09 05:10:21 ip-172-31-33-167 kubearmor[376347]: 2023-09-09 05:10:21.332505        INFO        Updated 1 security rule(s) to container_namespace/fuga/kubearmor_fuga
Sep 09 05:10:21 ip-172-31-33-167 kubearmor[376347]: i: 0, EndPointName: hoge, number of policies: 1
Sep 09 05:10:21 ip-172-31-33-167 kubearmor[376347]: i: 1, EndPointName: fuga, number of policies: 1
```

---

### Test case 2

1. Create  three policy files (`block_diff.yaml`, `block_ls.yaml`, `block_sleep.yaml`)

<details>
<summary> block_diff.yaml, block_ls.yaml, and block_sleep.yaml  </summary>

```yaml
ubuntu ~
> cat block_diff.yaml
apiVersion: security.kubearmor.com/v1
kind: KubeArmorPolicy
metadata:
  name: block-diff
spec:
  selector:
    matchLabels:
      # policyName
      kubearmor.io/container.name: hoge
  process:
    matchPaths:
      - path: /usr/bin/diff
  action:
    Block
ubuntu ~
> cat block_ls.yaml
apiVersion: security.kubearmor.com/v1
kind: KubeArmorPolicy
metadata:
  name: block-ls
spec:
  selector:
    matchLabels:
      # policyName
      kubearmor.io/container.name: hoge
  process:
    matchPaths:
      - path: /usr/bin/ls
  action:
    Block
ubuntu ~
> cat block_sleep.yaml
apiVersion: security.kubearmor.com/v1
kind: KubeArmorPolicy
metadata:
  name: block-sleep
spec:
  selector:
    matchLabels:
      # policyName
      kubearmor.io/container.name: hoge
  process:
    matchPaths:
      - path: /usr/bin/sleep
  action:
    Block
```

</details>

2. apply `block_diff.yaml` and `block_ls.yaml`

3. `dm.EndPoints` are shown below.

```bash
Sep 09 05:16:39 ip-172-31-33-167 kubearmor[377481]: 2023-09-09 05:16:39.844018        INFO        Detected a Container Security Policy (added/container_namespace/block-diff)
Sep 09 05:16:39 ip-172-31-33-167 kubearmor[377481]: 2023-09-09 05:16:39.867626        INFO        Registered the AppArmor profile (kubearmor_hoge)
Sep 09 05:16:39 ip-172-31-33-167 kubearmor[377481]: 2023-09-09 05:16:39.893401        INFO        Updated 1 security rule(s) to container_namespace/hoge/kubearmor_hoge
Sep 09 05:16:39 ip-172-31-33-167 kubearmor[377481]: i: 0, EndPointName: hoge, number of policies: 1


Sep 09 05:16:48 ip-172-31-33-167 kubearmor[377481]: 2023-09-09 05:16:48.835678        INFO        Detected a Container Security Policy (added/container_namespace/block-ls)
Sep 09 05:16:48 ip-172-31-33-167 kubearmor[377481]: 2023-09-09 05:16:48.862113        INFO        Updated 2 security rule(s) to container_namespace/hoge/kubearmor_hoge
Sep 09 05:16:48 ip-172-31-33-167 kubearmor[377481]: i: 0, EndPointName: hoge, number of policies: 2
```

4. Update below `kubearmor.io/container.name` of `block_diff.yaml`.

The container name in `block_diff.yaml` was changed from `hoge` to `fuga`, and apply the policy.

5. `dm.EndPoints` are shown below.

```bash
Sep 09 05:18:01 ip-172-31-33-167 kubearmor[377481]: 2023-09-09 05:18:01.761718        INFO        Detected a Container Security Policy (added/container_namespace/block-diff)
Sep 09 05:18:01 ip-172-31-33-167 kubearmor[377481]: 2023-09-09 05:18:01.785688        INFO        Registered the AppArmor profile (kubearmor_fuga)
Sep 09 05:18:01 ip-172-31-33-167 kubearmor[377481]: 2023-09-09 05:18:01.811592        INFO        Updated 1 security rule(s) to container_namespace/fuga/kubearmor_fuga
Sep 09 05:18:01 ip-172-31-33-167 kubearmor[377481]: i: 0, EndPointName: hoge, number of policies: 1
Sep 09 05:18:01 ip-172-31-33-167 kubearmor[377481]: i: 1, EndPointName: fuga, number of policies: 1
```

4. Update `kubearmor.io/container.name` of `block_ls.yaml`.

The container name in `block_ls.yaml` was changed from `hoge` to `fuga`, and apply the policy.

5. `dm.EndPoints` are shown below.

```bash
Sep 09 05:18:20 ip-172-31-33-167 kubearmor[377481]: 2023-09-09 05:18:20.732560        INFO        Detected a Container Security Policy (added/container_namespace/block-ls)
Sep 09 05:18:20 ip-172-31-33-167 kubearmor[377481]: 2023-09-09 05:18:20.759344        INFO        Updated 2 security rule(s) to container_namespace/fuga/kubearmor_fuga
Sep 09 05:18:20 ip-172-31-33-167 kubearmor[377481]: i: 0, EndPointName: fuga, number of policies: 2
```

6. apply `block_sleep.yaml`

7. `dm.EndPoints` are shown below.

```bash
Sep 09 05:18:44 ip-172-31-33-167 kubearmor[377481]: 2023-09-09 05:18:44.382272        INFO        Detected a Container Security Policy (added/container_namespace/block-sleep)
Sep 09 05:18:44 ip-172-31-33-167 kubearmor[377481]: 2023-09-09 05:18:44.408475        INFO        Updated 1 security rule(s) to container_namespace/hoge/kubearmor_hoge
Sep 09 05:18:44 ip-172-31-33-167 kubearmor[377481]: i: 0, EndPointName: fuga, number of policies: 2
Sep 09 05:18:44 ip-172-31-33-167 kubearmor[377481]: i: 1, EndPointName: hoge, number of policies: 1
```

8. Update `kubearmor.io/container.name` of `block_sleep.yaml`.

The container name in `block_sleep.yaml` was changed from `hoge` to `fuga`, and apply the policy.

9. `dm.EndPoints` are shown below.

```bash
Sep 09 05:19:03 ip-172-31-33-167 kubearmor[377481]: 2023-09-09 05:19:03.104947        INFO        Detected a Container Security Policy (added/container_namespace/block-sleep)
Sep 09 05:19:03 ip-172-31-33-167 kubearmor[377481]: 2023-09-09 05:19:03.131804        INFO        Updated 3 security rule(s) to container_namespace/fuga/kubearmor_fuga
Sep 09 05:19:03 ip-172-31-33-167 kubearmor[377481]: i: 0, EndPointName: fuga, number of policies: 3
```

From all results, the policy doesn't apply to the previous container, and apply to the new container.

**Additional information for reviewer?** :
_Mention if this PR is part of any design or a continuation of previous PRs_


**Checklist:**
- [x] Bug fix. Fixes #1387
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [x] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Commit has unit tests
- [ ] Commit has integration tests

<!--

The PR title message must follow convention:
`<type>(<scope>): <subject>`.

Where: <br />
- `type` is to define what type of PR is this.
  Most common types are:
    - `feat`      - for new features, not a new feature for build script
    - `fix`       - for bug fixes or improvements, not a fix for build script
    - `chore`     - changes not related to production code
    - `docs`      - changes related to documentation
    - `style`     - formatting, missing semi colons, linting fix etc; no significant production code changes
    - `test`      - adding missing tests, refactoring tests; no production code change
    - `refactor`  - refactoring production code, eg. renaming a variable or function name, there should not be any significant production code changes

- `scope` is a single word that best describes where the changes fit.
    - feature(`monitor`,`enforcer`)
    - test(`tests`, `bdd`)
    - chore(`build`)
- `subject` is a single line brief description of the changes made in the pull request.

-->

daemon1024 reviewed (approved):
LGTM 🙌 Thanks a lot for the PR. We are in the process of fixing our CI we will start merging PRs then.

Aryan-sharma11 reviewed (changes requested):
@haytok  Thanks for the PR, I tried it locally but still the policy is being enforced on the previous container as well. Imo simply updating the endpoint won't work as we need to update the container rules for the previous container too.
You can also test it locally by executing in the previous and the new container and running the commands.
For instance I have taken wordpress docker image from [here](https://github.com/docker/awesome-compose/tree/master/wordpress-mysql) 
and applied the Policy 
```
apiVersion: security.kubearmor.com/v1
kind: KubeArmorPolicy
metadata:
  name: ksp-block-policy
spec:
  severity: 3
  selector:
    matchLabels:
      kubearmor.io/container.name: wordpress-mysql-wordpress-1
  process:
    matchPaths:
    - path: /usr/bin/apt
    - path: /usr/bin/apt-get

      # apt update
      # apt-get update

  action:
    Block
```
-try executing in the contanier and run  commands `apt and apt update` they should be denied

- And now update the container name to  `wordpress-mysql-db-1` 
- Apply the policy and run the commands by  executing in the containers.  It is being denied in both the containers.  This should not be the case instead enforcement should be done only on `wordpress-mysql-db-1`

cc:- @daemon1024

Aryan-sharma11 commented on KubeArmor/core/unorchestratedUpdates.go:
```diff
@@ -503,6 +523,12 @@ func (dm *KubeArmorDaemon) ParseAndUpdateContainerSecurityPolicy(event tp.K8sKub
 			dm.EndPoints = append(dm.EndPoints, newPoint)
 		} else {
+			for idx, endPoint := range dm.EndPoints {
+				if kl.MatchIdentities(secPolicy.Spec.Selector.Identities, endPoint.Identities) {
```
Is there a reason to do this step again, as we have already updated value of `i` above at [line](https://github.com/kubearmor/KubeArmor/pull/1406/files#diff-168b945c99cdc3669f72461c0f4913020872a0e7ff4222676b356125676a4c64R431)?

haytok replied:
This is because if the `dm.Endpoints` object is deleted by the following process, the value of `i`, which indicates the index of the object, has changed.

```
dm.EndPoints = append(dm.EndPoints[:idx], dm.EndPoints[idx+1:]...)
```

https://github.com/kubearmor/KubeArmor/pull/1406/files#diff-168b945c99cdc3669f72461c0f4913020872a0e7ff4222676b356125676a4c64R443


If i is not recalculated, an out-of-array reference error will occur in the following operations.

```
dm.endpoints [i] = NewPoint
```

https://github.com/kubearmor/KubeArmor/pull/1406/files#diff-168b945c99cdc3669f72461c0f4913020872a0e7ff4222676b356125676a4c64R532

Aryan-sharma11 replied:
🤔 That makes sense. So can we assign the value of i after [line 458](https://github.com/kubearmor/KubeArmor/pull/1406/files#diff-168b945c99cdc3669f72461c0f4913020872a0e7ff4222676b356125676a4c64R458)
and remove the loop from [here 431](https://github.com/kubearmor/KubeArmor/pull/1406/files#diff-168b945c99cdc3669f72461c0f4913020872a0e7ff4222676b356125676a4c64R431) 
this would also remove the need to reassign the value to `i` from [here 526 ]( https://github.com/kubearmor/KubeArmor/pull/1406/files#diff-168b945c99cdc3669f72461c0f4913020872a0e7ff4222676b356125676a4c64R526).

Aryan-sharma11 commented on KubeArmor/core/unorchestratedUpdates.go:
```diff
@@ -436,6 +436,26 @@ func (dm *KubeArmorDaemon) ParseAndUpdateContainerSecurityPolicy(event tp.K8sKub
 		}
 	}
 
```
please add a comment specifying why are we doing this operation.

haytok replied:
What do you think of the comments below?

```
	// When the container name (value of kubearmor.io/container.name) specified in KubeArmorPolicy is changed, it is necessary to update the policies for the container with that name.
	for idx, endPoint := range dm.EndPoints {
		for policyIndex, policy := range endPoint.SecurityPolicies {
			if policy.Metadata["namespaceName"] == secPolicy.Metadata["namespaceName"] && policy.Metadata["policyName"] == secPolicy.Metadata["policyName"] && endPoint.EndPointName != containername {
```

Aryan-sharma11 replied:
how about 
``` 
// Updating container rules if there exists another container with same  `policy.Metadata["policyName"]` 
```

Aryan-sharma11 reviewed (commented):
Thanks @haytok, I have tested it locally, and the changes are working as expected.  Just some small things to be addressed.

Aryan-sharma11 commented on KubeArmor/core/unorchestratedUpdates.go:
```diff
@@ -429,12 +429,59 @@ func (dm *KubeArmorDaemon) ParseAndUpdateContainerSecurityPolicy(event tp.K8sKub
+					dm.RuntimeEnforcer.UpdateSecurityPolicies(endPoint)
+				}
+			}
+		} else {
```
Imo, we don't need this else statement wdyt ?

Aryan-sharma11 reviewed (commented):
@haytok Is there a specific reason to change the approach here as your earlier implementation was pretty straightforward imo? Also can you please attach the outputs?

daemon1024 commented on KubeArmor/core/unorchestratedUpdates.go:
```diff
@@ -429,12 +429,59 @@ func (dm *KubeArmorDaemon) ParseAndUpdateContainerSecurityPolicy(event tp.K8sKub
+		for policyIndex, policy := range endPoint.SecurityPolicies {
+			if policy.Metadata["namespaceName"] == secPolicy.Metadata["namespaceName"] && policy.Metadata["policyName"] == secPolicy.Metadata["policyName"] && endPoint.EndPointName != containername {
+				renamedContainerIndex = idx
+				deletedPolicyIndex = policyIndex
```
Today KubeArmor only supports, single container policies. But in future there will be Policy Matching to suppport policies on multiple containers. This logic will not scale when we add that. We should implement this in such a way, that it's not a special case handling but handle it in a natural way so as to address the issue. WDYT?

daemon1024 commented on KubeArmor/core/unorchestratedUpdates.go:
```diff
@@ -426,15 +426,70 @@ func (dm *KubeArmorDaemon) ParseAndUpdateContainerSecurityPolicy(event tp.K8sKub
+					EndPointName:       endPoint.EndPointName,
+					DeletedPolicyIndex: policyIndex,
+				}
+				renamedContainers = append(renamedContainers, renamedContainer)
```
Any particular reason to store it in an array and loop over it again? Why not just call the deletion code here itself?

Aryan-sharma11 commented on KubeArmor/core/unorchestratedUpdates.go:
```diff
@@ -426,15 +426,50 @@ func (dm *KubeArmorDaemon) ParseAndUpdateContainerSecurityPolicy(event tp.K8sKub
 	for idx, endPoint := range dm.EndPoints {
-		if kl.MatchIdentities(secPolicy.Spec.Selector.Identities, endPoint.Identities) {
-			i = idx
+		endPointIndex += 1
```
please handle this lint warning

haytok committed:
fix(core): fix to prevent policies from being applied to containers that shouldn't be applied

When running KubeArmor in systemd mode, when the container name (the value
of kubearmor.io/container.name) in the applied policy was changed and the
policy was applied, there was an issue where that policy was applied not
only to the new container but also to the previous container.

This bug was also reported in issue/#1387.

Therefore, in this commit, the policy doesn't apply to the previous
container, and apply to the new container.

Signed-off-by: Hayato Kiwata <haytok@amazon.co.jp>
